felix/bpf/maps.go

```diff
@@ -314,15 +337,99 @@ func (b *PinnedMap) Open() error {
 	return err
 }
 
+// nolint
+func (b *PinnedMap) pinnedMapMatchesConfiguration(maxEntries int) bool {
+	return maxEntries == b.MaxEntries
+}
+
+// nolint
+func (b *PinnedMap) migratePinnedMap(from, to string) error {
+	err := RepinMap(b.versionedName(), to)
+	if err != nil {
+		return fmt.Errorf("error repinning %s to %s: %w", from, to, err)
+	}
+	err = os.Remove(from)
+	if err != nil {
+		return fmt.Errorf("error removing the pin %s", from)
+	}
+	return nil
+}
+
+func (b *PinnedMap) oldMapExists() bool {
+	_, err := os.Stat(b.Path() + "_old")
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
 func (b *PinnedMap) EnsureExists() error {
+	var oldfd MapFD
 	if b.fdLoaded {
 		return nil
 	}
 
+	// In case felix restarts in the middle of migration, we might end up with
+	// old map. If the old map size and config matches, pin the old map back.
+	// open the fd and return
+	if b.oldMapExists() {
+		// try to open the old map
+		fd, err := GetMapFDByPin(b.versionedFilename() + "_old")
+		if err != nil {
+			return fmt.Errorf("error opening old map %s, err = %w", b.versionedFilename()+"_old", err)
+		}
+
+		defer fd.Close()
+		// Get the old map info
+		mapInfo, err := GetMapInfo(fd)
+		if err != nil {
+			return fmt.Errorf("error getting map info of the old map %w", err)
+		}
+
+		// If the old map's size matches the config, repin old to
+		// the actual path, try to open.
+		if b.pinnedMapMatchesConfiguration(mapInfo.MaxEntries) {
+			err = b.migratePinnedMap(b.Path()+"_old", b.Path())
+			if err != nil {
+				return fmt.Errorf("error pinning old map %w", err)
+			}
+		} else {
+			return fmt.Errorf("config does not match old map size, adjust the config")
```

Comment:
My expectation was that felix would restore the old map as the current and try to apply the configuration again as that might be a good configuration already adjusted by the user.

If that is not the desired flow, we should tell the user to what value to adjust the config - what the expected value is.

Reply:
Agreed. If there is a old map, it makes sense to repin to the actual path and let it continue. Modified the code to do this.